judge/models/tests/test_contest.py

```diff
@@ -0,0 +1,454 @@
+        contest.format_config = {}
+        with self.assertRaisesRegex(ValidationError, 'Contest problem label script'):
+            contest.full_clean()
+        contest.problem_label_script = '''
+            function(n)
+                return n
+            end
+        '''
+        # Test for caching
+        with self.assertRaisesRegex(ValidationError, 'Contest problem label script'):
+            contest.full_clean()
+        del contest.get_label_for_problem
+        with self.assertRaisesRegex(ValidationError, 'should return a string'):
+            contest.full_clean()
+        contest.problem_label_script = ''
+        del contest.get_label_for_problem
+        contest.full_clean()
+
+    def test_normal_user_current_contest(self):
+        current_contest = self.profiles['normal'].current_contest
+        self.assertIsNotNone(current_contest)
+
+        current_contest.set_disqualified(True)
+        self.profiles['normal'].refresh_from_db()
+        self.assertTrue(current_contest.is_disqualified)
+        self.assertIsNone(self.profiles['normal'].current_contest)
+        self.assertEqual(current_contest.score, -9999)
+
+        current_contest.set_disqualified(False)
+        self.profiles['normal'].refresh_from_db()
+        self.assertFalse(current_contest.is_disqualified)
+        self.assertIsNone(self.profiles['normal'].current_contest)
+        self.assertEqual(current_contest.score, 0)
+
+    def test_live_participation(self):
+        participation = ContestParticipation.objects.get(
+            contest=self.hidden_scoreboard_contest,
+            user=self.profiles['normal'],
+            virtual=ContestParticipation.LIVE,
+        )
+        self.assertTrue(participation.live)
+        self.assertFalse(participation.spectate)
+        self.assertEqual(participation.end_time, participation.contest.end_time)
+        self.assertFalse(participation.ended)
+        self.assertIsInstance(participation.time_remaining, timezone.timedelta)
+
+    def test_spectating_participation(self):
+        participation = create_contest_participation(
+            contest='hidden_scoreboard',
+            user='superuser',
+            virtual=ContestParticipation.SPECTATE,
+        )
+
+        self.assertFalse(participation.live)
+        self.assertTrue(participation.spectate)
+        self.assertEqual(participation.start, participation.contest.start_time)
+        self.assertEqual(participation.end_time, participation.contest.end_time)
+
+    def test_virtual_participation(self):
+        participation = create_contest_participation(
+            contest='private',
+            user='superuser',
+            virtual=1,
+        )
+
+        self.assertFalse(participation.live)
+        self.assertFalse(participation.spectate)
+        self.assertEqual(participation.start, participation.real_start)
+        self.assertIsInstance(participation.end_time, timezone.datetime)
+
+
+class ContestTagTestCase(TestCase):
```

Comment:
Should this test that `text_color` caches its values and uses them on repeated runs?

Reply:
I'd actually like to know this for other tests as well (such as utils).

Reply:
From some discussion on Slack, we decided that it's not worth to test this, as some Python hacks would be needed to grab the `cache` from the `test_color` property.